File: src/breslin/math/vector_3d.php
var Vector3D = new Class(
{
	
initialize: function()
{
	this.x = 0;
	this.y = 0;
	this.z = 0;
},

log: function(msg)
{
        setTimeout(function()
        {
                throw new Error(msg);
        }, 0);
},

printValues: function()
{
	this.log('xx: ' + this.x + 'yy: ' + this.y + 'zz: ' + this.z); 	
},

returnValues: function()
{
	return 'xx: ' + this.x + 'yy: ' + this.y + 'zz: ' + this.z; 	
},

length: function()
{
        return Math.sqrt(this.x*this.x + this.z*this.z);
},

isZero: function()
{
        if (this.x == 0.0 && this.y == 0.0 && this.z == 0.0)
        {
                return true;
        }
        else
        {
                return false;
        }
	
},

zero: function()
{
	this.x = 0
	this.y = 0
	this.z = 0
},

normalise: function()
{
	len = this.length();
	if (len == 0)
	{
		return;
	}
	this.x /= len;
	this.y = 0;
	this.z /= len;
},

parseFloat: function()
{
	this.x = parseFloat(this.x);
	this.y = parseFloat(this.y);
	this.z = parseFloat(this.z);
},

//multiply this vector by a scalar
multiply: function(num)
{
	this.x = parseFloat(this.x) * num;
	this.y = parseFloat(this.y) * num;
	this.z = parseFloat(this.z) * num;
},

add: function(v)
{
	x = parseFloat(this.x) + parseFloat(v.x);
	y = 0;
	z = parseFloat(this.z) + parseFloat(v.z);


	this.x = parseFloat(x);
	this.z = parseFloat(z);
},

subtract: function(vectorToSubtract)
{
        this.x = this.x - vectorToSubtract.x;
        this.y = this.y - vectorToSubtract.y;
        this.z = this.z - vectorToSubtract.z;
},

copyValuesFrom: function(copyFrom)
{
	this.x = copyFrom.x;
	this.y = copyFrom.y;
	this.z = copyFrom.z;
},

dot: function(v2)
{
	d = this.x * v2.x + this.y * v2.y + this.z * v2.z;
}

});

File: src/breslin/math/vector3D.cpp
#include "vector3D.h"

//log
#include "../clientside/tdreamsock/dreamSockLog.h"

#include <math.h>

Vector3D::Vector3D()
{
	x = 0;
	y = 0;
	z = 0;

	mQuaternion = new Quaternion; 
}

Vector3D::Vector3D(float x1, float y1, float z1)
{
	x = x1;
	y = y1;
	z = z1;
}

Vector3D::~Vector3D()
{
}

void Vector3D::truncate(float max)
{
	if (this->length() > max)
  	{
    		this->normalise();
    		this->multiply(max);
  	} 	
}

float Vector3D::length()
{
	return sqrt(x*x + y*y + z*z);
}

bool Vector3D::isZero()
{
	if (x == 0.0 && y == 0.0 && z == 0.0)
	{
		return true;
	}
	else 
	{
		return false;
	}
}

void Vector3D::printValues()
{
	LogString("x:%f",x);
	LogString("z:%f",z);
}

void Vector3D::zero()
{
	x = 0;
	y = 0;
	z = 0;
}

void Vector3D::normalise()
{
	float len = length();
	if (len == 0)
		return;
	x /= len;
	y /= len;
	z /= len;
}

//multiply this vector by a scalar
void Vector3D::multiply(float num)
{
	x = x * num;
	y = y * num;
	z = z * num;
}

//add another vector to this one
void Vector3D::add(Vector3D* vectorToAddtoThisOne)
{
	x = x + vectorToAddtoThisOne->x;
	y = y + vectorToAddtoThisOne->y;
	z = z + vectorToAddtoThisOne->z;
}

void Vector3D::add(Vector3D* add1, Vector3D* add2)
{
	x = add1->x - add2->x;
	y = add1->y - add2->y;
	z = add1->z - add2->z;	
}

//subtract another vector from this one
void Vector3D::subtract(Vector3D* vectorToSubtract)
{
	x = x - vectorToSubtract->x;
	y = y - vectorToSubtract->y;
	z = z - vectorToSubtract->z;
}

void Vector3D::subtract(Vector3D* sub1, Vector3D* sub2)
{
	x = sub1->x - sub2->x;
	y = sub1->y - sub2->y;
	z = sub1->z - sub2->z;	
}

//copy values
void Vector3D::copyValuesFrom(Vector3D* copyFrom)
{
	x = copyFrom->x;
	y = copyFrom->y;
	z = copyFrom->z;
}

float Vector3D::dot(Vector3D* v2)
{
	float d = x * v2->x + y * v2->y + z * v2->z;
	return d;
}


Vector3 Vector3D::convertToVector3()
{
	Vector3 vector3;
	vector3.x = x;
	vector3.y = y;
	vector3.z = z;

	return vector3;
}


Quaternion* Vector3D::getRotationTo(Vector3D* to)
{
	Vector3D fallbackAxis;

	Vector3D v0;
	Vector3D v1;
	v0.copyValuesFrom(this);
	v1.copyValuesFrom(to);

	v0.normalise();	
	v1.normalise();

	float d = v0.dot(&v1);

    	// If dot == 1, vectors are the same
    	if (d >= 1.0f)
    	{
		mQuaternion = new Quaternion(1.0,0.0,0.0,0.0);
		return mQuaternion;
    	}
			
	if (d < (1e-6f - 1.0f))
	{
		// rotate 180 degrees about the fallback axis
		Vector3 fb;
		fb.x = fallbackAxis.x;
		fb.y = fallbackAxis.y;
		fb.z = fallbackAxis.z;
		
		// rotate 180 degrees about the fallback axis
		mQuaternion->FromAngleAxis(Radian(Math::PI), fb);
	}
	else
	{
		Real s = Math::Sqrt( (1+d)*2 );
        	Real invs = 1 / s;

		Vector3D c;
		c = v0.crossProduct(v1);

   	    	mQuaternion->x = c.x * invs;
       		mQuaternion->y = c.y * invs;
        	mQuaternion->z = c.z * invs;
        	mQuaternion->w = s * 0.5f;
		mQuaternion->normalise();
	}
	return mQuaternion;
}

Vector3D Vector3D::crossProduct(Vector3D b)
{
	Vector3D c;
	c.x = this->y * b.z - this->z * b.x;
	c.y = this->z * b.x - this->x * b.z;
	c.z = this->x * b.y - this->y * b.x;

	return c;
}


//calculate how far off we are from some vector
float Vector3D::getDegreesToSomething(Vector3D* to)
{
	mQuaternion = this->getRotationTo(to);
	
    	// convert to degrees
    	Real degreesToSomething = mQuaternion->getYaw().valueDegrees();
	return degreesToSomething;
}

Vector3 Vector3D::getVector3()
{
	Vector3 vector3;
	vector3.x = x;
	vector3.y = y;
	vector3.z = z;
	return vector3;
}

void Vector3D::convertFromVector3(Vector3 vector3)
{
	x = vector3.x;
	y = vector3.y;
	z = vector3.z;

}
Vector3D Vector3D::getVectorOffset(float offset, bool degrees)
{
	//get the sine and cosine of 90degrees

	double cs;
	double sn;

	if (degrees)
	{
        	cs = cos( offset * 3.14f / 180.0f);
        	sn = sin( offset * 3.14f / 180.0f);
	}
	else
	{
        	cs = cos(offset);
        	sn = sin(offset);
	}

        Vector3D offsetVelocity;

        offsetVelocity.x = this->x * cs - this->z * sn;
        offsetVelocity.z = this->x * sn + this->z * cs;

	return offsetVelocity;
}

